Add GPU based sample laser sensor to simulate Livox mid360 sensor

georgflick opened the pull request:
This PR adds the following:
- commits from the main gazebo repo: [https://github.com/gazebosim/gazebo-classic](https://github.com/gazebosim/gazebo-classic) to keep our devel-enway up-to-date
- commits from Jacobs PR from the cube #3205

My changes:
  - Added a new protobuf type `laserscan_angles`, which extends the type `laserscan` with the `azimuth` and `zenith` angles
  - To use the sample_sensor, you need to set the parameters:
    - `sample_sensor` to true
    - `sample_size` to define the amount of "samples", which are published using the angles in the csv files for each step
    - `sample_csv_file` to a csv file with the angles for the sample sensor
    - to reduce the gpu load while simulating multiple sensors, the texture image was changed from float32 RGB values to float16 R values. Using only one float16 value per pixel, it is not possible to simulate the intensity. The intensity is then set to a fixed value.

How-To-Test:
- Test together with PR in main repo

Roboterbastler committed:
Re-implement GpuLaser with cubemap to lift vFOV limit

Roboterbastler committed:
Add back GpuLaser::cameraCount member

Roboterbastler committed:
Add assertion for laser scan material pointer

georgflick committed:
Add laserscan message with angles

georgflick committed:
Add sample offset to iterate through sample list

georgflick committed:
Limit camera resolution to 1024 pixel for sample sensor

georgflick committed:
Save and restore the camera of the gpu laser instead of only rotating

instead of rotate the camera back and forth.

Roboterbastler commented on gazebo/rendering/GpuLaser.cc:
```diff
@@ -427,59 +439,67 @@ GpuLaser::DataIter GpuLaser::LaserDataBegin() const
 {
   const unsigned int index = 0;
   // Data stuffed into three floats (RGB)
-  const unsigned int skip = 3;
+  unsigned int skip = 3;
```
This can be constexpr right?

Roboterbastler replied:
Actually the other const variables could also be made constexpr here

Roboterbastler commented on gazebo/rendering/GpuLaser.cc:
```diff
@@ -427,59 +439,67 @@ GpuLaser::DataIter GpuLaser::LaserDataBegin() const
   // range data in R channel
   const unsigned int rangeOffset = 0;
   // intensity data in G channel
   const unsigned int intenOffset = 1;
-  return DataIter(index, this->dataPtr->laserBuffer.data(), skip, rangeOffset,
-      intenOffset, this->dataPtr->w2nd);
+  return {index, this->dataPtr->laserBuffer.data(), skip, rangeOffset,
+          intenOffset, this->dataPtr->horizontalRangeCount};
 }
 
 //////////////////////////////////////////////////
 GpuLaser::DataIter GpuLaser::LaserDataEnd() const
 {
-  const unsigned int index = this->dataPtr->h2nd * this->dataPtr->w2nd;
+
+  unsigned int index;
+  index = this->dataPtr->verticalRangeCount * this->dataPtr->horizontalRangeCount;
 
   // Data stuffed into three floats (RGB)
-  const unsigned int skip = 3;
+  unsigned int skip = 3;
```
Similar here, both can be const/constexpr, no need to assign `index` in next line.

Roboterbastler commented on gazebo/sensors/GpuRaySensor.cc:
```diff
@@ -698,73 +787,152 @@ bool GpuRaySensor::UpdateImpl(const bool /*_force*/)
-    scan->set_ranges(i, range);
-    scan->set_intensities(i, intensity);
-  }
+      // ToDo Georg replace!
```
Left over todo comment?

georgflick replied:
Removed

Roboterbastler commented on gazebo/rendering/GpuLaser.cc:
```diff
@@ -427,59 +439,67 @@ GpuLaser::DataIter GpuLaser::LaserDataBegin() const
+          intenOffset, this->dataPtr->horizontalRangeCount};
+}
+
+//////////////////////////////////////////////////
+std::vector<float>* GpuLaser::LaserData() const
+{
+  return &this->dataPtr->laserBuffer;
 }
```
Return const reference to vector instead of pointer

georgflick replied:
Done

Roboterbastler commented on gazebo/sensors/GpuRaySensor.cc:
```diff
@@ -698,73 +787,152 @@ bool GpuRaySensor::UpdateImpl(const bool /*_force*/)
-  }
+      // ToDo Georg replace!
+      std::vector<float>* data = this->dataPtr->laserCam->LaserData();
+      constexpr double fixed_intensity = 1.0;
```
Maybe it would make sense to make this a configurable parameter?

georgflick replied:
Done, you can now set the parameter `fixed_intensity` and if the parameter `compute_intensity` is set to `false`, the intensity is set to `fixed_intensity`.

Roboterbastler reviewed (changes requested):
Wouldn't it make sense to make it a general option to disable the intensity configuration? The motivation is purely performance related right?
Maybe it would also be beneficial for other (non-sample) sensors if this could be disabled and we save a lot of computation (time). I think we don't use the intensities much, if we even use them anywhere at all...

Roboterbastler commented:
Should we drop the existing commits concerning the cubemap implementation from `devel-enway`? Maybe it would be much cleaner if the commits in this PR replace those.

Roboterbastler commented on gazebo/sensors/GpuRaySensor.cc:
```diff
@@ -804,6 +972,69 @@ double GpuRaySensor::NextRequiredTimestamp() const
+//////////////////////////////////////////////////
+bool GpuRaySensor::readCsvFile(std::string file_name, std::vector<std::vector<double>>& datas)
+{
+  std::fstream file_stream;
+  file_stream.open(file_name, std::ios::in);
+  if (file_stream.is_open()) {
+      std::string header;
+      std::getline(file_stream, header, '\n');
+      while (!file_stream.eof()) {
+          std::string line_str;
+          std::getline(file_stream, line_str, '\n');
+          std::stringstream line_stream;
+          line_stream << line_str;
+          std::vector<double> data;
+          try {
+              while (!line_stream.eof()) {
+                  std::string value;
+                  std::getline(line_stream, value, ',');
+                  data.push_back(std::stod(value));
+              }
+          } catch (...) {
+              continue;
+          }
+          datas.push_back(data);
+      }
+      gzdbg << "data size:" << datas.size() << "\n";
+      return true;
+  } else {
+      gzerr << "cannot read csv file!" << file_name << "\n";
+  }
+  return false;
+}
```
Seems to be copied from livox plugin, maybe add notice to license on top to be on the safe side?

georgflick replied:
Ah right, the copy of the function was more or less for testing. I replaced it now with my own function to load the csv file.

georgflick committed:
Add parameter compute_intensity and fixed_intensity

disable compute_intensity to save some performance.
If it is disabled, the intensity is set to the fixed value automatically

georgflick commented:
> Wouldn't it make sense to make it a general option to disable the intensity configuration? The motivation is purely performance related right? Maybe it would also be beneficial for other (non-sample) sensors if this could be disabled and we save a lot of computation (time). I think we don't use the intensities much, if we even use them anywhere at all...

True, I added a parameter `compute_intensity` , if that is off the texture is only `R_FLOAT16` and intensity is set to the parameter `fixed_intensity`.

Roboterbastler reviewed (approved):
I tested it again and looks good!

georgflick commented:
Because of the merging conflicts, from updating the commits to the newest version 11.2.0 from the upstream gazebo-classic repo, I will push the gazebo11 PR branch manually to `devel-enway` and merge the PR afterwards.
